### ISSUE TITLE
reliability issue with vehicles

### ISSUE THREAD

**Issue by BL97** [OpenLoco/OpenLoco#1689]
- Platform: win 10
 - Version: 22.10 win 32
 - Build/commit: ?

vehicle reliability is lower than should be on new vehicles

purchase a vehicle, for example the re 460 locomotive (with cheats) pre 1991, notice reliability 100%
purchase the same vehicle in 1991, notice reliability 77%
purchase again in 1993, notice reliability still 77%

purchasing the vehicle in 1970 for example, with cheats, will allow you to see the true reliability (100%)

in vanilla when a vehicle is unlocked it has a reliability penalty for the first 2 years, the re 460 for example has a reliability of 77% when first unlocked in 1991, however after 2 years the penalty goes away meaning if the vehicle is purchased again in 1993 the reliability will be 100%, this applies to all vehicles in game iirc

**Comment by duncanspumpkin** [comment]
So this is the current code for reliability:
```cpp
        int32_t reliability = vehObject.reliability * 256;
        if (getCurrentYear() + 2 > vehObject.designed)
        {
            // Reduce reliability by an eighth after 2 years past design
            reliability -= reliability / 8;
            if (getCurrentYear() + 3 > vehObject.designed)
            {
                // Reduce reliability by a further eighth (quarter total) after 3 years past design
                reliability -= reliability / 8;
            }
        }
        if (reliability != 0)
        {
            reliability += 255;
        }
```

from looking at the assembly again its a bit different:
```cpp
        int32_t reliability = vehObject.reliability * 256;
        if (vehObject.designed + 2 > getCurrentYear())
        {
            reliability -= reliability / 8;
            if (vehObject.designed + 3 != getCurrentYear())
            {
                reliability -= reliability / 8;
            }
        }
        if (reliability != 0)
        {
            reliability += 255;
        }
```

**Comment by duncanspumpkin** [comment]
Lets walk through a scenario with the vanilla assembly. Assume the vehicle was designed in 1990 and has a reliability of 75%. We want to see what its reliability is at 1989, 1990, 1991, 1992, 1993, 1994.

Ignore the *256 and the += 255 they are just so that maths works in integers.

1989: 1990 + 2 > 1989... yes, 75 - 9.... = 66 1990 + 3 != 1989... yes, 66 - 8 = 58.
1990: 1990 + 2 > 1990... yes, 75 - 9.... = 66 1990 + 3 != 1990... yes, 66 - 8 = 58.
1991: 1990 + 2 > 1991... yes, 75 - 9.... = 66 1990 + 3 != 1991... yes, 66 - 8 = 58.
1992: 1990 + 2 > 1992... no, 75.
1993: 1990 + 2 > 1993... no, 75.
... 75.
So it looks like there is a penalty for the first 2 years but then just a flat value. 

To me this all looks like a bug in the original as the `if (vehObject.designed + 3 != getCurrentYear())` will never happen. So then the question becomes what behaviour do we want. Vanilla? What we currently have? Something different? It seems the intention was that there would be a period of time where there was a penalty and as that time increased the penalty would increase. But perhaps we could do something a bit different like a 2 year penalty then a number of years bonus followed by a final penalty.

**Comment by Dracheron** [comment]
I'm trying to think of something new, but give me some time to figure out the exact details on it.  Also, the idea of 100% reliability before something is invented seems absurd as well in general.

In general, I was thinking a logistical increase as it comes online and readily available, a slight logistical drop to deal with actual vehicle age, and a logistical drop as it becomes obsolete it will become less reliable as well (note that vehicles that do not become obsolete won't have that third portion since it won't make any sense to have it)

I also think a few variables in determining the obsolete factor would also make things interesting (i.e. North America will always have some parts available for example, where as the two european areas might lose it completely after a certain point)

so in part, you have an initial logistical incline as it becomes available, a slight logistical decrease with age, and an obsolete logistical decrease partially determined by location.

Anyone got any input as to the idea I have?

**Comment by duncanspumpkin** [comment]
Although it would be nice to link it to regions I'd ideally want to keep this self contained and without requiring any new state. A logarithmic drop is the obvious candidate for a falloff probably want it to come down to a fixed value rather than zero. Building things before invented is only ever possible with cheats so I'm personally less concerned what happens there.

**Comment by Dracheron** [comment]
yeah, though I think a logistic function is probably what's more appropriate for both aging and the vehicles going obsolete (if a vehicle isn't going to go obsolete it won't have that function inherently)

And as the building before invented thing yeah even though its only possible doing it using cheats it would be nice to consider prototyping, etc. into that calculation

I can see this needing to be self contained as well

The other solution is that each region has its own equation for something going obsolete

**Comment by Dracheron** [comment]
Here's what I'm starting to think: the reliability of the vehicles will be taken as a ceiling sum of four different factors.

The First function is effectively a population growth function handling reliability of once its invented.  This is taken at maximum theoretical value if it does not have an invention year.
The Second function is an uneven distribution with a peak near the invention year to deal with parts availability.  For vehicles that were invented earlier than 1895 or have no invention year this is instead replaced with the second half of a bell curve.  If there is no obsolescence year, the function is instead replaced with a second population growth curve.
The third function is an arctangent decay function with age of the vehicle
The fourth function is a logistic decay function with obsolescence of the vehicle.  This is also taken at theoretical maximum value if there is no obsolescence year.

The biggest question I'm left with for developing these various functions is at what interval is reliability calculated at so that the calculation is not left hanging for values and that equations can be made consistent with each other.

**Comment by duncanspumpkin** [comment]
That sounds a bit complex. This is for the initial reliability for a bought train. Keep in mind all objects will have a field from the object creator that is a reliability value 0-100. This value should be an input into this function. The function doesn't get called often at all so we can make it complex if we want. Why don't you draw some graphs of how you would expect it to work. Model a few objects with different reliability inputs and at different years from design.

**Comment by DC-4444** [comment]
> So this is the current code for reliability:
> 
>         int32_t reliability = vehObject.reliability * 256;
>         if (getCurrentYear() + 2 > vehObject.designed)
>         {
>             // Reduce reliability by an eighth after 2 years past design
>             reliability -= reliability / 8;
>             if (getCurrentYear() + 3 > vehObject.designed)
>             {
>                 // Reduce reliability by a further eighth (quarter total) after 3 years past design
>                 reliability -= reliability / 8;
>             }
>         }
>         if (reliability != 0)
>         {
>             reliability += 255;
>         }
> from looking at the assembly again its a bit different:
> 
>         int32_t reliability = vehObject.reliability * 256;
>         if (vehObject.designed + 2 > getCurrentYear())
>         {
>             reliability -= reliability / 8;
>             if (vehObject.designed + 3 != getCurrentYear())
>             {
>                 reliability -= reliability / 8;
>             }
>         }
>         if (reliability != 0)
>         {
>             reliability += 255;
>         }

Seems that in the current code, the variables in the if statements are in the wrong order. Since current year + 2 and current year + 3 will always be greater than the year the vehicle is designed in, the if statements will always be true. Because of this, the reliability penalty will always be applied no matter what year it is after the vehicle is designed instead of only for the first two years.

To work the same as in vanilla, the first if statement would be:
`if vehObject.designed + 2 > getCurrentYear()`

And the second if statement would be:
`if vehObject.designed + 3 > getCurrentYear()`
